corehq/apps/repeaters/tasks.py

```diff
@@ -17,75 +21,40 @@
 )
 def check_repeaters():
     start = datetime.utcnow()
-    LIMIT = 100
-    progress_report = new_progress_report()
-
-    def loop():
-        # take LIMIT records off the top
-        # the assumption is that they all get 'popped' in the for loop
-        # the only exception I can see is if there's a problem with the
-        # locking, a large number of locked tasks could pile up at the top,
-        # so make a provision for that worst case
-        number_locked = progress_report['number_locked']
-        repeat_records = RepeatRecord.all(
-            due_before=start,
-            limit=LIMIT + number_locked
-        )
-        return process_repeater_list(
-            repeat_records,
-            start=start,
-            cutoff=start + CHECK_REPEATERS_INTERVAL,
-            progress_report=progress_report
-        )
-
-    while loop():
-        pass
-
-    now = datetime.utcnow()
-    progress_report['timedelta'] = unicode(now - start)
-    progress_report['time'] = unicode(now)
-    logging.info(json.dumps(progress_report))
+    cutoff = start + CHECK_REPEATERS_INTERVAL
 
+    redis_client = get_redis_client().client.get_client()
 
-def new_progress_report():
-    return {'success': [], 'fail': [], 'locked': [], 'deleted': [], 'number_locked': 0}
-
+    for record in iterate_repeat_records(start):
+        now = datetime.utcnow()
+        lock_key = _get_repeat_record_lock_key(record)
 
-def process_repeater_list(repeat_records, start=None, cutoff=datetime.max, progress_report=None):
-    """
-    Attempts to process a list of repeat records, with a bunch of optional arguments that are
-    used by the periodic task.
-    """
-    start = start or datetime.utcnow()
-    progress_report = progress_report or new_progress_report()
-    DELETED = '-Deleted'
+        if now > cutoff:
```

Comment:
i think we can still end up with a situation where many of these tasks run at the same time, for example, if the queue gets backed up and they all start at the same time, then they'd all run together for about 5 minutes.  that doesn't seem like it would do too much harm since we lock the call to process_repeat_record.delay, but it could end up clogging the periodic queue.  if we think it's not a big deal then probably fine to leave as is, but it might be worth it to have this task also try to acquire a separate non-blocking lock that expires in 4 minutes 50 seconds and if it can't acquire the lock it just returns.